Create analyzer in ComInterfaceGenerator to enforce the generator is used with IUnknown only

jtschuster opened the pull request:
Creates an analyzer that warns if the GeneratedComInterfaceAttribute is
used with InterfaceTypeAttribute that has an argument that is not
'InterfaceIsIUnknown'

jtschuster committed:
Create analyzer to enforce Generator is used with IUnknown only

Creates an analyzer that warns if the GeneratedComInterfaceAttribute is
used with InterfaceTypeAttribute that has an argument that is not
'InterfaceIsIUnknown'

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratorDiagnostics.cs:
```diff
@@ -21,6 +20,7 @@ public class Ids
             public const string InvalidLibraryImportAttributeUsage = Prefix + "1050";
             public const string TypeNotSupported = Prefix + "1051";
             public const string ConfigurationNotSupported = Prefix + "1052";
+            public const string InterfaceTypeNotSupported = "1053";
```
```suggestion
            public const string InterfaceTypeNotSupported = Prefix + "1053";
```

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,96 @@
+    {
+        private const string _generatedComInterfaceAttributeName = "GeneratedComInterfaceAttribute";
+        private const string _generatedComInterfaceAttributeNamespace = "System.Runtime.InteropServices.Marshalling";
+#if DEBUG
+        private const string _generatedComInterfaceAttributeAssemblyName = "Microsoft.Interop.Ancillary";
+#else
+        private const string _generatedComInterfaceAttributeAssemblyName = "System.Runtime.InteropServices";
+#endif
```
We can just match the attribute based on name. We don't need to match on the symbol.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,96 @@
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            if (!Debugger.IsAttached)
+                context.EnableConcurrentExecution();
```
We should always enable concurrent execution for the analyzer. If that makes the development experience worse, we can update the test harness to disable concurrent execution for all of our analyzer tests when the debugger is attached (which wouldn't be that bad of an idea).

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/GeneratedComInterfaceAnalyzerTests.cs:
```diff
@@ -0,0 +1,185 @@
+            #pragma warning restore CS8019
+            """;
+
+        public static async Task<ImmutableArray<Diagnostic>> CompileAndRunAnalyzer(params string[] codeSnippets)
+        {
+            Compilation comp = await TestUtils.CreateCompilation(codeSnippets);
+            // Ensure no compiler errors
+            Assert.Empty(comp.GetDiagnostics());
+            // Ensure no analyzer erros
+            return await TestUtils.RunAnalyzers(comp, new GeneratedComInterfaceAttributeAnalyzer());
+        }
```
Instead of creating your own compilation and running the analyzers, we should use the Roslyn Analyzer Test SDK here as we do for the ConvertToLibraryImportAnalyzerTests test suite. That test will handle validating all of the diagnostic locations and ids at a much higher fidelity than just checking the number of diagnostics.

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratorDiagnostics.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

jtschuster committed:
PR Feedback

- Use test harness from LibraryImportGenerator.Unit.Tests
- Don't check for assembly name when matching attribute
- Undo changes from TestUtils

jtschuster committed:
Undo auto formatting

jtschuster committed:
Checkout System.Private.CoreLib.Shared.projitems from upstream/main

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/ComInterfaceGenerator.Unit.Tests.csproj:
```diff
@@ -19,6 +19,12 @@
             Link="Common\CustomStructMarshallingCodeSnippets.cs" />
     <Compile Include="..\Common\ICustomMarshallingSignatureTestProvider.cs"
             Link="ICustomMarshallingSignatureTestProvider.cs" />
+    <Compile Include="..\LibraryImportGenerator.UnitTests\Verifiers\CSharpAnalyzerVerifier.cs"
+            Link="Verifiers\CSharpAnalyzerVerifier.cs"/>
+    <Compile Include="..\LibraryImportGenerator.UnitTests\Verifiers\CSharpCodeFixVerifier.cs"
+            Link="Verifiers\CSharpCodeFixVerifier.cs"/>
+    <Compile Include="..\LibraryImportGenerator.UnitTests\Verifiers\CSharpVerifierHelper.cs"
+            Link="Verifiers\CSharpVerifierHelper.cs"/>
```
Let's move these 3 files to the `Common` folder that `TestUtils.cs` is in now that we're using them from multiple unit test projects.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/GeneratedComInterfaceAnalyzerTests.cs:
```diff
@@ -0,0 +1,195 @@
+            #pragma warning restore CS8019
+            """;
+
+        public static async Task<ImmutableArray<Diagnostic>> CompileAndRunAnalyzer(params string[] codeSnippets)
```
Looks like we can remove this function.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/GeneratedComInterfaceAnalyzerTests.cs:
```diff
@@ -0,0 +1,195 @@
+
+                """;
+            await VerifyCS.VerifyAnalyzerAsync(_usings + snippet);
+
```
```suggestion
```

jtschuster committed:
Move verifiers to common directory and remove unused method

jtschuster committed:
Remove empty line

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratorDiagnostics.cs:
```diff
@@ -21,6 +20,7 @@ public class Ids
             public const string InvalidLibraryImportAttributeUsage = Prefix + "1050";
             public const string TypeNotSupported = Prefix + "1051";
             public const string ConfigurationNotSupported = Prefix + "1052";
+            public const string InterfaceTypeNotSupported = Prefix + "1053";
```
Are we expecting other diagnostics around invalid usage of `GeneratedComInterfaceAttribute`? If so, I think it would make sense to have this ID represent invalid usage of the attribute (similar to what we did for `LibraryImport` with `SYSLIB1050`) rather than be specific to `InterfaceType`.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class GeneratedComInterfaceAttributeAnalyzer : DiagnosticAnalyzer
+    {
+        private const string _generatedComInterfaceAttributeName = "GeneratedComInterfaceAttribute";
+        private const string _generatedComInterfaceAttributeNamespace = "System.Runtime.InteropServices.Marshalling";
+
+        private const string _interfaceTypeAttributeName = "InterfaceTypeAttribute";
+        private const string _interfaceTypeAttributeNamespace = "System.Runtime.InteropServices";
+
```
Might make sense to start a file with constants like these. I imagine we'll need them elsewhere.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+#nullable enable
```
Do we need to explicitly enable nullable? I thought we had it enabled in the project as a whole.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+namespace Microsoft.Interop
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class GeneratedComInterfaceAttributeAnalyzer : DiagnosticAnalyzer
```
Should we put analyzers in an Analyzers subfolder and namespace? That is what we did for `LibraryImportGenerator` and I liked the separation/organization.

jtschuster replied:
I like that too, I've move the related files to an Analyzers/ folder

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratorDiagnostics.cs:
```diff
@@ -145,6 +145,16 @@ public class Ids
                 isEnabledByDefault: true,
                 description: GetResourceString(nameof(SR.ConfigurationNotSupportedDescription)));
 
+        public static readonly DiagnosticDescriptor InterfaceTypeNotSupported =
```
Do we want this here or in a separate file specifically for analyzer diagnostics (as opposed to generator diagnostics)? We kept them separate for `LibraryImportGenerator`. I could see keeping the `DiagnosticDescriptor` separate, but merging the IDs into one file so that it is easy to see them all / which values are used in one place. Either way, we should make both the interop generators follow the same pattern.

jtschuster replied:
I've moved them to their own file in the Analyzers/ folder.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
I wouldn't expect this to run on generated code - is there a specific reason it needs to?

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+            {
+                INamedTypeSymbol typeSymbol = (INamedTypeSymbol)context.Symbol;
+                if (typeSymbol.TypeKind != TypeKind.Interface)
+                    return;
```
```suggestion
                    return;

```

elinor-fung commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,87 @@
+                    return;
+
+                // Interfaces should not have both GeneratedComInterfaceTypeAttribute and InterfaceTypeAttribute
+                if (GetGeneratedComInterfaceAttribute(typeSymbol, out AttributeData? generatedComInterfaceAttribute)
```
```suggestion
                if (GetGeneratedComInterfaceAttribute(typeSymbol, out AttributeData? _)
```
We don't seem to actually be using it (at least right now).

CyrusNajmabadi replied:
can also just do `out _` fwiw :)

elinor-fung reviewed (commented):
What is the expected behaviour of the generator if it encounters this case? Will it just generate code for `IUnknown` / completely ignore `InterfaceType`? If we might put the user in an unexpectedly bad state, would it make sense to be an error diagnostic in the generator instead?

jkoritzinsky commented:
> What is the expected behaviour of the generator if it encounters this case? Will it just generate code for `IUnknown` / completely ignore `InterfaceType`? If we might put the user in an unexpectedly bad state, would it make sense to be an error diagnostic in the generator instead?

The `InterfaceType` attribute basically allows developers to specify the "base" interface type for their COM interface in the built-in system. Users can pick, `IUnknown`, `IDispatch`, or `IInspectable` (there's also a `Dual` option which means `IDispatch` + actually generate vtable members for the methods).

This attribute would allow users to explicitly state that they want an `IUnknown`-based interface, which will be our default with the source generator and would block users from trying to say they want a different base interface that we don't support.

elinor-fung commented:
If a project has an interface with `[GeneratedComInterface(...)]` and `[InterfaceType(ComInterfaceType.IInspectable)]`, but has analyzers turned off and uses the generator, what is the user's experience? Would the generator:
1. Just generate what we support, effectively ignoring `InterfaceType`
    - This would leave the user in a bad/unsupported state
2. Actually block/fail
    - If the generator does this, I don't know that we need an analyzer to also emit a diagnostic

I know we hit the 'analyzers disabled, generator enabled' case with `LibraryImport` and moved our invalid attribute usage diagnostics to the generator instead of analyzers.

jkoritzinsky commented:
I think the generator would just skip generating what is not supported.

I know that @CyrusNajmabadi was pushing for diagnostics like these to be moved to analyzers for better performance in the IDE. I'd be curious to hear his thoughts on the "analyzers disabled, generator enabled" scenario.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx

Co-authored-by: Elinor Fung <elfung@microsoft.com>

jtschuster committed:
Merge branch 'GenComIntAnalyzer' of https://github.com/jtschuster/runtime into GenComIntAnalyzer

jtschuster committed:
PR feedback

Move Analyzer and related diagnostics to Analayzers/ folder
Change diagnostic to "InvalidUse of GeneratedComInterfaceAttribute"
Add diagnostic to list-of-diagnostics.md
Add and organize tests using all variations of ComInterfaceType

jtschuster commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratorDiagnostics.cs:
```diff
@@ -21,6 +20,7 @@ public class Ids
             public const string InvalidLibraryImportAttributeUsage = Prefix + "1050";
             public const string TypeNotSupported = Prefix + "1051";
             public const string ConfigurationNotSupported = Prefix + "1052";
+            public const string InterfaceTypeNotSupported = Prefix + "1053";
```
I've changed it to 1090, which was the first unused diagnostic I saw.

jtschuster commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Resources/Strings.resx:
```diff
@@ -185,4 +185,10 @@
+    <value>Speficied InterfaceType is not supported with GeneratedComInterfaceAttribute</value>
+  </data>
+  <data name="InterfaceTypeNotSupportedMessage" xml:space="preserve">
+    <value>Using GeneratedComInterfaceAttribute with InterfaceTypeAttribute is only supported with argument InterfaceIsIUnknown</value>
```
Updated the message to include the unsupported value.

jtschuster committed:
Update list-of-diagnostics.md

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Disable tests on Mono

jtschuster committed:
Add ActiveIssue to all test classes

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Skip tests on browser

jtschuster committed:
Revert "Skip tests on browser" - failures on browser were unrelated

This reverts commit 8375b91ace1233c9b5bd7dbd527a4609afbc121f.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into GenComIntAnalyzer

jtschuster committed:
Merge branch 'GenComIntAnalyzer' of https://github.com/jtschuster/runtime into GenComIntAnalyzer

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/GeneratedComInterfaceAnalyzerTests.cs:
```diff
@@ -0,0 +1,758 @@
+        }
+
+    [ActiveIssue("https://github.com/dotnet/runtime/issues/60650", TestRuntimes.Mono)]
+        public class InterfaceHasGeneratedComInterfaceAttributeOnly
```
All of the tests in this class are the same since the `InterfaceTypeAttribute` is not included. Can we collapse these down to one test for `[GeneratedComInterface]` without any `InterfaceTypeAttribute`?

jtschuster committed:
Remove redundant tests

jkoritzinsky commented:
Failure was a known issue with our usage of homebrew. This is ready to be merged.
